server/services/property.service.js

```diff
@@ -525,3 +525,96 @@ export const deleteFloorPlan = async (floorPlanDetails) => {
     throw new ErrorHandler(httpStatus.BAD_REQUEST, 'Error deleting floor plan', error);
   }
 };
+
+export const addImage = async (imageDetails) => {
+  const property = await getPropertyById(imageDetails.propertyId).catch((error) => {
+    throw new ErrorHandler(httpStatus.INTERNAL_SERVER_ERROR, 'Internal Server Error', error);
+  });
+
+  if (!property) {
+    throw new ErrorHandler(httpStatus.NOT_FOUND, 'Invalid property');
+  }
+
+  if (imageDetails.vendorId.toString() !== property.addedBy.toString()) {
+    throw new ErrorHandler(httpStatus.FORBIDDEN, 'You are not permitted to perform this action');
+  }
+
+  let { gallery } = property;
+
+  gallery = gallery.filter((image) => image.title === imageDetails.title);
+
+  if (gallery.length > 0) {
+    throw new ErrorHandler(httpStatus.PRECONDITION_FAILED, 'Image with title already exists');
+  }
+
+  try {
+    return Property.findByIdAndUpdate(
+      property._id,
+      { $push: { gallery: { title: imageDetails.title, url: imageDetails.url } } },
+      { new: true },
+    );
+  } catch (error) {
+    throw new ErrorHandler(httpStatus.BAD_REQUEST, 'Error addding image', error);
+  }
+};
+
+export const updateImage = async (updatedImage) => {
+  const property = await getPropertyById(updatedImage.propertyId).catch((error) => {
+    throw new ErrorHandler(httpStatus.INTERNAL_SERVER_ERROR, 'Internal Server Error', error);
+  });
+
+  if (!property) {
+    throw new ErrorHandler(httpStatus.NOT_FOUND, 'Invalid property');
+  }
+
+  if (updatedImage.vendorId.toString() !== property.addedBy.toString()) {
+    throw new ErrorHandler(httpStatus.FORBIDDEN, 'You are not permitted to perform this action');
+  }
+
+  let { gallery } = property;
+
+  gallery = gallery.filter((image) => image.title === updatedImage.title);
+
+  if (gallery.length > 0 && gallery[0]._id.toString() !== updatedImage.imageId.toString()) {
```

Comment:
I think we can check for the imageId in the filter above
```suggestion
  if (gallery.length > 0) {
```